### PR TITLE
[WIP, Feature] Structured hydra config: Loss

### PR DESCRIPTION
**Pull request by GavinPHR**
Hydra configs for loss of different models.
Will do a merge from base branch, after exploration PR #556 is merged.

### CODE REVIEW

File: test/test_configs.py [hunk 1]
```diff
@@ -516,6 +517,151 @@ def test_exploration(self, network, wrapper):
         value(tensordict)
 
 
+from torchrl.modules import (
```
Comment (vmoens) [inline test/test_configs.py:520]: why do we do the import here?

File: test/test_configs.py [hunk 2]
```diff
@@ -442,6 +442,21 @@ def test_redq(self, pixels, independent, continuous):
         qvalue(tensordict)
 
 
+def make_halfcheetah_env_with_state_transforms():
+    import torchrl.envs.transforms as T
```
Comment (vmoens) [inline test/test_configs.py:446]: let's avoid doing imports in functions!

File: test/test_configs.py [hunk 3]
```diff
@@ -516,6 +517,151 @@ def test_exploration(self, network, wrapper):
...
+    NormalParamWrapper,
+    ValueOperator,
+    DdpgMlpActor,
+    DdpgMlpQNet,
+)
+
+
+@pytest.mark.skipif(not _has_hydra, reason="No hydra found")
+class TestLossConfigs:
+    import torchrl.objectives.costs as costs
```
Comment (vmoens) [inline test/test_configs.py:534]: let's avoid importing stuff everywhere

File: test/test_configs.py [hunk 4]
```diff
@@ -516,6 +517,151 @@ def test_exploration(self, network, wrapper):
...
+
+    def _make_actor_dqn(self, env):
+        network = DuelingMlpDQNet(
+            out_features=list(env.action_spec.shape),
```
Comment (vmoens) [inline test/test_configs.py:538]: why not `env.action_spec.shape[-1]`?

File: test/test_configs.py [hunk 5]
```diff
@@ -516,6 +517,151 @@ def test_exploration(self, network, wrapper):
...
+            spec=env.action_spec,
+            distribution_class="TanhNormal",
+        )
+        qvalue_operator = ValueOperator(
+            MLP(num_cells=[10], out_features=1),
+            in_keys=["action", "observation_vector"],
+        )
+        return policy_operator, qvalue_operator
+
+    @pytest.mark.parametrize("model", ["ddpg", "dqn", "ppo", "redq", "sac"])
```
Comment (vmoens) [inline test/test_configs.py:627]: this is a bit too general
we should have one function per loss, and test the hyperparams of each of them independently

**Review by vmoens — COMMENTED** [review]
We're on the right track with this! Great stuff

File: test/test_configs.py [hunk 6]
```diff
@@ -516,6 +526,208 @@ def test_exploration(self, network, wrapper):
...
+    def test_ppo_loss(self):
+        config = ["loss=ppo_loss"]
+        cfg = hydra.compose("config", overrides=config)
+        env = make_halfcheetah_env_with_state_transforms()
+        loss_partial = instantiate(cfg.loss)
+        actor, critic = self._make_model_ppo(env)
+        critic(actor(env.reset()))
+        loss = loss_partial(actor, critic)
+
+        assert type(loss) is getattr(torchrl.objectives.costs, "PPOLoss")
```
Comment (vmoens) [inline test/test_configs.py:657]: see SACLoss comment

File: test/test_configs.py [hunk 7]
```diff
@@ -516,6 +526,208 @@ def test_exploration(self, network, wrapper):
...
+    def test_dqn_loss(self):
+        config = ["loss=dqn_loss"]
+        cfg = hydra.compose("config", overrides=config)
+        env = make_halfcheetah_env_with_state_transforms()
+        loss_partial = instantiate(cfg.loss)
+        actor = self._make_actor_dqn(env)
+        actor(env.reset())
+        loss = loss_partial(actor)
+
+        assert type(loss) is getattr(torchrl.objectives.costs, "DQNLoss")
```
Comment (vmoens) [inline test/test_configs.py:644]: see SACLoss comment

File: test/test_configs.py [hunk 8]
```diff
@@ -516,6 +526,208 @@ def test_exploration(self, network, wrapper):
...
+    def test_ddpg_loss(self):
+        config = ["loss=ddpg_loss"]
+        cfg = hydra.compose("config", overrides=config)
+        env = make_halfcheetah_env_with_state_transforms()
+        loss_partial = instantiate(cfg.loss)
+        actor, qvalue = self._make_model_ddpg(env)
+        qvalue(actor(env.reset()))
+        loss = loss_partial(actor, qvalue)
+
+        assert type(loss) is getattr(torchrl.objectives.costs, "DDPGLoss")
```
Comment (vmoens) [inline test/test_configs.py:631]: see SACLoss comment

File: test/test_configs.py [hunk 9]
```diff
@@ -516,6 +526,208 @@ def test_exploration(self, network, wrapper):
...
+    def test_redq_loss(self):
+        config = ["loss=redq_loss"]
+        cfg = hydra.compose("config", overrides=config)
+        env = make_halfcheetah_env_with_state_transforms()
+        loss_partial = instantiate(cfg.loss)
+        actor, qvalue = self._make_model_redq(env)
+        qvalue(actor(env.reset()))
+        loss = loss_partial(actor, qvalue)
+
+        assert type(loss) is getattr(torchrl.objectives.costs, "REDQLoss")
```
Comment (vmoens) [inline test/test_configs.py:682]: see SACLoss comment

**Review by vmoens — APPROVED** [review]
Great work! 
Can you just check the type assertions?

File: test/test_configs.py [hunk 10]
```diff
@@ -516,6 +526,208 @@ def test_exploration(self, network, wrapper):
...
+    def test_sac_loss(self):
+        config = ["loss=sac_loss"]
+        cfg = hydra.compose("config", overrides=config)
+        env = make_halfcheetah_env_with_state_transforms()
+        loss_partial = instantiate(cfg.loss)
+        actor, qvalue, value = self._make_model_sac(env)
+        value(qvalue(actor(env.reset())))
+        loss = loss_partial(actor, qvalue, value)
+
+        assert type(loss) is getattr(torchrl.objectives.costs, "SACLoss")
```
[the remark this answers is not in the document]
Comment (GavinPHR) [reply test/test_configs.py]: I didn't know why I used `getattr()`, should all be changed to `isinstance()` now.
